### DISCUSSION TITLE
Defining the Actors and Roles - who is the Audience?

### DISCUSSION THREAD

**Discussion by taylor** [lfn-cnti/bestpractices#30] [General]
The actors (entity playing a specific role) is important to provide context for any best practice proposed. It's important that the CNF WG have a shared definition of each actor.   This first description will be updated periodically. Please see the comments for more frequent updates. After there is enough consensus through discussion we will add the information to a more "permanent home" in the repository.

---

**Groups/orgs**
- Service Provider (SP)
- Telco Vendors
- SPs Customer/Consumer
- Open source projects/initiatives (eg. [TIP](https://telecominfraproject.com/), [ONF](https://opennetworking.org/), [Kamailio](https://www.kamailio.org/w/)/[OpenSIPS](https://opensips.org/), [Litmus](https://github.com/litmuschaos/litmus))
- Other Network Operators (eg. enterprise)

**Telecom Actors:**
- Architect - Responsible for defining the HW and SW components utilized by the SP. Uses modeling tools to lay out the service
- CNF Developer - Creates telco software
   - Infrastructure
   - Application (aka workload) 
      > could be the same or different team/group/org
- Operation (Ops) Person
    - Application/workload
    - Infrastructure
    - Network
    - Security
      > could be the same or different team/group/org
- System Integrator

**Roles**
- 1
- 2

The same or similar actors can be found across the Teleco domain: eg. there may be Infra / app developers in a Vendor org, in an open source project, or at an SP. It's important to communicate where the actor is as their motivations in addition to any shared across groups.

When the context is not immediately apparent by other explicit informations the actors should be combined with the group and any other information. Examples:
- Service Provider Architect
- Service Provider Security Ops Person
- Vendor CNF Infrastructure Developer

**Comment by jeffsaelens** [comment]
Maybe we make Ops teams a subset of SP, and add a second subset architecture/development? Also, and this might be what operations was, but maybe change that to network operator? as a distinction for a SP specifically.

**Reply by jeffsaelens** [reply]
it was brought to my attention I was not clear thinking something like:

- SP
  - Architecture/Development
  - Operations

- Network Operators (enterprise)

and then the rest the same for now (to include operations in both SP and Ent having the domains you listed). This suggestion is not perfect in its current form, just want the other big network operators that don't necessarily have the specific SP challenges those of us on the provider side do are represented.

**Reply by taylor** [reply]
Addressed by having actors outside of the groups.

**Comment by taylor** [comment]
How about this:

- Service Provider (SP)
  - Architect
  - Developer
  - Operation (Ops) person (combined or split)
    - Application/workload
    - Infrastructure
    - Network
    - Security
- CNF Developers - (eg. Vendors)
   - Infrastructure
   - Application (aka workload) 
      > could be the same or different team/group/org
- SPs Customer/Consumer
- Other Network Operators (eg. enterprise)


Note: the sub groups would be combined with their higher level actor, like:
- Service Provider Architect
- SP Ops Person
- CNF Infrastructure Developer

**Reply by jeffsaelens** [reply]
Looks good.

**Reply by electrocucaracha** [reply]
I kind of confused about the differences between a SP Developer and SP Ops Application/workload. 

In another hand, SP Consumer is not the same as CNF Customer? Maybe I associate SP to Public Cloud providers

**Reply by jeffsaelens** [reply]
SP developer could be axed but its someone who writes code full time in an SP. Think AirShip Devs in ATT. SP Ops App people are operations folks that operate apps from internal devs and vendor devs.

SP customer is someone who consumes services from the SP directly.

**Reply by taylor** [reply]
@jeffsaelens could we say that a AirShip Developer at ATT is the same as an AirShip Developer at Cisco and likewise a SP CNF (Infra/App) Developer is the same as a Vendor CNF (Infra/App) Developer if they are building the same type of software?

Based on that I would update the CNF Developer section like this:

CNF Developers -  (eg. SPs, Vendors, Other Network Operators)
- Infrastructure
- Application (aka workload)

**Reply by taylor** [reply]
@electrocucaracha, the SP is the customer of the CNF Developer so we don't need to list it separately, but we could explicitly communicate that when we fully define what each of the actors are.

**Reply by electrocucaracha** [reply]
You have mention a couple of examples, should be better to include them as part of the definition?

**Reply by taylor** [reply]
@electrocucaracha which examples and where do you mean?   I have not updated the first post as of yet to get more feedback.  What do you suggest though?

**Reply by electrocucaracha** [reply]
> SP developer could be axed but its someone who writes code full time in an SP. Think AirShip Devs in ATT. SP Ops App people are operations folks that operate apps from internal devs and vendor devs.
> 
> SP customer is someone who consumes services from the SP directly.

So SP Ops App is not similar to a Cloud Administrator

**Reply by electrocucaracha** [reply]
> @electrocucaracha which examples and where do you mean? I have not updated the first post as of yet to get more feedback. What do you suggest though?

Or at least to expand the definition, like:

* Service Provider:
  - Architect: Responsible for defining the HW and SW components utilized by the SP
  - Developer: Creator of APIs to consume resources provided by the SP
  - Operator: Responsible to maintain up time of services provided.

**Reply by taylor** [reply]
@electrocucaracha a "Cloud Administrator" may be a someone who makes sure the infrastructure and its services work for applications running on it.  It could be a someone that understands how to manage and deploy applications on the infrastructure vs the Cloud App Developer doing the deployment and management.

The Service Provider will have an Ops folks with many responsibilities including taking care of the
- infrastructure
- networking (including WAN)
- security (network, infra and application)
- application life cycle

That could be multiple team or one team with mixed roles.  (Probably the former in the majority of cases)

**Reply by michaelspedersen** [reply]
For CNF Developers, would it make sense to include something like "Operation"? For scaling and managing there's going to be a need for telemetry and analytics which I see as being something that is provided by the CNF Dev/Vendor, as they have a better view of the capabilities of the workload compared to SP and more general Ops teams.

**Reply by rannyh** [reply]
I feel a bit uncomfortable with the top level of the tree being things like SP, Vendor, Customer, etc. It implies that there are different rules you should follow based on who's paying your salary. I am afraid this will perpetuate the adversarial approach that existed between SPs and vendors in the past. If you are a developer of apps, I don't think it matters if you work for a SP, Vendor, ISV or whatever. At the end of the day it is best if they all abide by the same rules that will make their app run properly on the cloud infrastructure. Operations personnel may also be employees of a vendor, in case of a BOT type of project. Again, their rules should not differ from the ones applicable to SP organics operations teams. 
So bottom line, I propose to have on the top level of the tree things like:
* Infrastructure developer
* Infrastructure operator
* Application developer
* Application operator
* Application consumer
* ...

**Reply by jeffsaelens** [reply]
I'm not sure if it comes across as different actors/different rules, but we'd definitely want to ensure that was not the case. The goal here should be the opposite, trying to come up with commonalities for all. The main driver I think @taylor was pushing for here is to show the that there are various actors, and each one would have unique motivations and needs to contribute to this effort. From there we could strive to bridge those gaps.

**Reply by taylor** [reply]
See top of discussion for updated list.  I've split the different groups from the actors.  The actors can be in any group. When talking about an actor it is probably a good idea to communicate what group they are in or explicitly communicate it is applicable to any group the actor is in.

**Reply by vukg** [reply]
Time to copy-paste from Slack here:

![image](https://user-images.githubusercontent.com/52979324/103752125-946c8600-5009-11eb-97ed-7ee9d06cb038.png)

**Reply by sronanrh** [reply]
Since a lot of conversations is around the fact that these practices will be used for RFP, do we need to add an Actor for CNF Purchaser as a consumer of the "certification"?

**Comment by kmoha-saleem** [comment]
Do we also need to consider the role of "System Integrators" or no?

**Reply by taylor** [reply]
Added to list. See top of discussion

**Comment by xmulligan** [comment]
This might be good to add to the introduction that @rabi-abdel is writing in #34  It might be helpful to talk about the unique motivations and needs of each user as @jeffsaelens said above

**Comment by iawells** [comment]
I think we need to take our user categories and lump them into responsibilities so we have fewer names to deal with in our user stories - whoever is integrating it has an integration role and it doesn't matter if they're a dedicated SI or an SP operator, it's more a question of what they do that matters (which we could define by example).

**Reply by jeffsaelens** [reply]
So just boil it down to the main 4?

Architect - Chief Power Point creator
Developer - Under appreciated martyr 
Systems Integrator - Person gluing the architects and developers insanity together
Operator - inheritor of the mess produce by the first three

Definitions/examples may need some tweaking

**Reply by rmerz** [reply]
I like the focus on the main 4. That would work for me.
I would somewhat complain about the powerpoint focus for architects, I mean it is really not just about powerpoint

Now, something else came to my mind: in any sort of transformation (and going CNF is certainly a transformation), you need management support. Isn't this another relevant actor?

**Comment by iawells** [comment]
I would focus on:

Roles:
- Network architect - planning out the network
- CNF dev - writes a component they would want to use
- Platform dev - writes an operable (i.e. deployable and upgradeable) platform
- SI - glues the CNF and platform network together, with configuration and either written processes or orchestration, per architect's design; and maybe patches up things the operator requests afterward
- Operator - day to day person that runs the resulting network

Then a section per role; and include in that section a discussion of the perspectives the role brings (e.g. reliability, security).

That's less fine grained than Taylor has in actors but the perspectives should cover the rest.

Then, organisations involved (the dirty reality):
- SP, platform vendor, telco vendor (the SP may not use vendors or they may), SI, ...
- Some example assignments of roles to orgs and additional considerations for that sort of breakdown, e.g.:
  -  communications between roles
  - purchasing
  - when CNF is bought in
  - implications of an RFP process)

This comes last because it's the one that diverts us from the more technical details.

**Comment by electrocucaracha** [comment]
These actors and roles have to be related with the [scope of this group](https://github.com/cncf/cnf-wg/tree/master/doc#scope)
